docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,265 @@
+When determining the destination for a given file, all sync rules will be considered.
+Thus, a single file may be copied to multiple destinations.
+
+Error cases:
+- It should be an error to specify `inferTo=true` together with `to` in the same sync rule.
+  ```yaml
+  - from: '*.js'
+    to: dest/
+    inferTo: true
+  ```
+- It should be an error to specify either of `flatten` or `strip` together with `inferTo`.
+  ```yaml
+  - from: '*.js'
+    inferTo: true
+    flatten: true        # ERR
+  - from: 'src/**/*.js'
+    inferTo: true
+    strip: src/          # ERR
+  ```
+- It should be an error to specify `strip` with a prefix which which does not match the static prefix of the `from` field.
+  ```yaml
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: src/sub        # ERR
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: app/           # ERR
+  ```
+
+Open questions about the configuration:
+
+**\<Is the syntax clear enough?\>**
+Is the syntax clear?
+Is the configuration intuitive to use?
+Is there any ambiguity which needs attention?
+
+Resolution: __Not Yet Resolved__
+
+**\<Does it improve clarity to have the `inferTo` field?\>**
+A different approach could be to fall back to inference, if the `to` field is left blank.
+This would have the advantage that there needs to be no validation about having `to` and `inferTo` in the same rule, which should cause an error.
+
+Resolution: __Not Yet Resolved__
+   
+**\<Error for `strip` and `flatten`\>**
+Given the new sync config, users may try to combine various flags in unintended ways.
+For example, should `strip` and `flatten` in the same rule always be an error?
+My intuition would say, that `strip+flatten` is the same as `flatten`, but this should to be discussed.
+
+Resolution: __Not Yet Resolved__
+   
+   
+#### Migration plan
+The new configuration supports all existing use-cases and is fully backwards compatible.
+An automatic schema upgrade shall be implemented.
+
+### Open Issues/Question
+
+**\<Overlapping copies\>** 
+Overlapping copies can be caused by at least two ways:
+
+1. The `COPY/ADD` instructions in the Dockerfile may copy to the same destination. For example:
+   ```dockerfile
+   COPY foo /foo
+   COPY bar/ /   # bar contains a file called "foo"
+   ```
+2. Two sync rules may copy files to the same destination. For example:
+   ```yaml
+   sync:
+   - from: foo
+     to: /baz
+   - from: bar
+     to: /baz
+   ```
+   
+Should such cases be treated as errors, warnings, or not be detected at all?
+
+Resolution: __Not Yet Resolved__
+
+**\<Transitive copies\>** 
+Multi-stage Dockerfiles can lead to quite contrived scenarios. For example:
+```dockerfile
+FROM scratch
+ADD foo baz
+FROM scratch
+COPY --from=0 baz bar
+```
+Ideally, the sync destination should be inferred as `foo -> bar`.
+
+This is quite an edge case and can easily be circumvented.
+However, it is quite complex to implement.
+Should we bother about this special case?
+   
+Resolution: __Not Yet Resolved__
```

Comment:
I would say, not right now. If people ask for it we can do a 2 step roll out where 
- first detect and warn
- then fail.